# Basic Exploratory Data Analysis (EDA)

This notebook contains a general EDA applied on the dataset

In [1]:
import pandas as pd
import numpy as np

In [2]:
from pathlib import Path

PROJECT_ROOT = Path.cwd()
DATA_DIR = PROJECT_ROOT / 'data'
BRONZE_DIR = DATA_DIR / 'bronze'
DATA_FILE = BRONZE_DIR / 'PS_20174392719_1491204439457_log.csv'

# Basic overview of the Data Structure

In [3]:
df = pd.read_csv(DATA_FILE)
df

,step,type,amount,nameOrig,oldbalanceOrg,newbalanceOrig,nameDest,oldbalanceDest,newbalanceDest,isFraud,isFlaggedFraud
0,1,PAYMENT,9839.64,C1231006815,170136.00,160296.36,M1979787155,0.00,0.00,0,0
1,1,PAYMENT,1864.28,C1666544295,21249.00,19384.72,M2044282225,0.00,0.00,0,0
2,1,TRANSFER,181.00,C1305486145,181.00,0.00,C553264065,0.00,0.00,1,0
3,1,CASH_OUT,181.00,C840083671,181.00,0.00,C38997010,21182.00,0.00,1,0
4,1,PAYMENT,11668.14,C2048537720,41554.00,29885.86,M1230701703,0.00,0.00,0,0
...,...,...,...,...,...,...,...,...,...,...,...
6362615,743,CASH_OUT,339682.13,C786484425,339682.13,0.00,C776919290,0.00,339682.13,1,0
6362616,743,TRANSFER,6311409.28,C1529008245,6311409.28,0.00,C1881841831,0.00,0.00,1,0
6362617,743,CASH_OUT,6311409.28,C1162922333,6311409.28,0.00,C1365125890,68488.84,6379898.11,1,0
6362618,743,TRANSFER,850002.52,C1685995037,850002.52,0.00,C2080388513,0.00,0.00,1,0


In [4]:
memory_usage = df.memory_usage(deep=True).sum()
print(f"Memory size: {memory_usage / (1024 ** 2):.2f} MB")

Memory size: 1598.19 MB


In [7]:
df.info(show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6362620 entries, 0 to 6362619
Data columns (total 11 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   step            6362620 non-null  int64  
 1   type            6362620 non-null  object 
 2   amount          6362620 non-null  float64
 3   nameOrig        6362620 non-null  object 
 4   oldbalanceOrg   6362620 non-null  float64
 5   newbalanceOrig  6362620 non-null  float64
 6   nameDest        6362620 non-null  object 
 7   oldbalanceDest  6362620 non-null  float64
 8   newbalanceDest  6362620 non-null  float64
 9   isFraud         6362620 non-null  int64  
 10  isFlaggedFraud  6362620 non-null  int64  
dtypes: float64(5), int64(3), object(3)
memory usage: 534.0+ MB


Observations: 
- 11 columns
- 6362620 rows
- No missing values
- dtypes: 3 int64, 3 object, 5 float64

In [8]:
df.describe()

,step,amount,oldbalanceOrg,newbalanceOrig,oldbalanceDest,newbalanceDest,isFraud,isFlaggedFraud
count,6.362620e+06,6.362620e+06,6.362620e+06,6.362620e+06,6.362620e+06,6.362620e+06,6.362620e+06,6.362620e+06
mean,2.433972e+02,1.798619e+05,8.338831e+05,8.551137e+05,1.100702e+06,1.224996e+06,1.290820e-03,2.514687e-06
std,1.423320e+02,6.038582e+05,2.888243e+06,2.924049e+06,3.399180e+06,3.674129e+06,3.590480e-02,1.585775e-03
min,1.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,1.560000e+02,1.338957e+04,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
50%,2.390000e+02,7.487194e+04,1.420800e+04,0.000000e+00,1.327057e+05,2.146614e+05,0.000000e+00,0.000000e+00
75%,3.350000e+02,2.087215e+05,1.073152e+05,1.442584e+05,9.430367e+05,1.111909e+06,0.000000e+00,0.000000e+00
max,7.430000e+02,9.244552e+07,5.958504e+07,4.958504e+07,3.560159e+08,3.561793e+08,1.000000e+00,1.000000e+00


Observations:

- Step: based on the mean and the std deviation, the distribution seems quite uniform.

- Amount: there is an enormous variation of amount of transferred money and the standard deviation is very high (603.858). 
50% of all transactions move below 74.872, 75% of all transactions move below 208.721 and the max amount is 92.445.200 so we will find very few but huge transactions (long tail).

- OldbalanceOrg, NewbalanceOrig, OldbalanceDest and NewbalanceDest: both have similar standard deviations and maximum values. It calls my attention that min, 25% and even 50% can be 0. That makes me wonder if those accounts have been opened or closed just to commit fraud. Some significant amount accounts have also completely completely emptied. 

- isFlaggedFraud: the mean is even lower but we have to take into consideration that this flag is activated only for transactions greater than 200.000. 



In [9]:
df.describe(include = ['O'])

,type,nameOrig,nameDest
count,6362620,6362620,6362620
unique,5,6353307,2722362
top,CASH_OUT,C1530544995,C1286084959
freq,2237500,3,113


Observations:

- Type: there are 5 types of transactions, the most frequent is CASH_OUT. We will analyse them and their relationship with fraudulent transactions.

- nameOrig and nameDest: some names are repeated. We will check if that has something to do with fraudulent behaviours. 

---
Deeper stats analysis will be performed in the following notebooks

### Constant Columns

In [10]:
constant_cols = []
cardinality = df.nunique()
constant_cols = cardinality[cardinality == 1].index.tolist()
if constant_cols:
    for col in constant_cols:
        constant_value = df[col].iloc[0] 
        print(f"'{col}': '{constant_value}'")
else:
    print("No constant columns detected")



No constant columns detected


### Cardinality

In [11]:
cardinality = df.nunique().sort_values(ascending=False)
print(cardinality.to_frame('Unique Values'))

                Unique Values
nameOrig              6353307
amount                5316900
oldbalanceDest        3614697
newbalanceDest        3555499
nameDest              2722362
newbalanceOrig        2682586
oldbalanceOrg         1845844
step                      743
type                        5
isFraud                     2
isFlaggedFraud              2


- nameOrig: Extremely high → probably a unique identifier for each sender (account ID). Drop or use for - aggregation
- amount: Continuous numeric feature. Maybe scale
- oldbalanceDest: Continuous numeric, possibly many zeros? Keep it
- newbalanceDest: Continuous numeric. Keep
- nameDest: High-cardinality categorical. Too many uniques → Drop, hash, or aggregate
- newbalanceOrig: Continuous numeric. Keep
- oldbalanceOrig: Continuous numeric. Keep
- step: Time step. Keep, can encode as time (sin/cos transformation or directly numeric)
- type: Transaction type. Keep OHT
- isFlaggedFraud: Binary target. Keep ofc

In [12]:
high_cardinality_cols = []
for col in df.select_dtypes(include='object').columns:
        if df[col].nunique() / len(df) > 0.95:
            high_cardinality_cols.append(f"'{col}' has very high cardinality ({df[col].nunique():,} unique values)")

for col in high_cardinality_cols:
    print(col)
print("-" *50)
print(f" The length of the dataframe is: {len(df):,}")

'nameOrig' has very high cardinality (6,353,307 unique values)
--------------------------------------------------
 The length of the dataframe is: 6,362,620


### Negative values 

In [13]:
neg_vals = []
for col in df.select_dtypes(include=[np.number]).columns:
    if (df[col] < 0).any():
        n_negative = (df[col] < 0).sum()
        neg_vals.append(f" This {col} has {n_negative:,} negative values")

if neg_vals:
    for val in neg_vals:
        print(neg_vals)
else:
    print("No obvious negative values detected")

No obvious negative values detected


### Outliers

In [14]:
for col in df.select_dtypes(include=[np.number]).columns:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    
    outliers = df[(df[col] < lower_bound) | (df[col] > upper_bound)]
    if len(outliers) > 0:
        pct = len(outliers) / len(df) * 100
        print(f"  {col}: {len(outliers):,} outliers ({pct:.2f}%)")

  step: 102,688 outliers (1.61%)
  amount: 338,078 outliers (5.31%)
  oldbalanceOrg: 1,112,507 outliers (17.49%)
  newbalanceOrig: 1,053,391 outliers (16.56%)
  oldbalanceDest: 786,135 outliers (12.36%)
  newbalanceDest: 738,527 outliers (11.61%)
  isFraud: 8,213 outliers (0.13%)
  isFlaggedFraud: 16 outliers (0.00%)


- step : probably not a real outlier
- amount: expected, transactions are usually heavy tailed
- old/new balances in Orig/Dest: also expected. Balances vary a lot across users
- isFlaggedFraud: the target

## Business consistency

In [16]:
data = df.copy()

inconsistent_cols = []
data['balance_diff'] = data['oldbalanceOrg'] - data['newbalanceOrig']

# check balance_diff and amount are equal, with allowed difference of |1|
inconsistent = data[~np.isclose(data['balance_diff'], data['amount'], atol=1.0)]
if len(inconsistent) > 0:
    pct = len(inconsistent) / len(data) * 100
    inconsistent_cols.append(f" {len(inconsistent):,} transactions ({pct:.2f}%) have inconsistent balance calculations")
    print(inconsistent_cols)
else:
    print(" Balance calculations are consistent")

[' 5,000,782 transactions (78.60%) have inconsistent balance calculations']


In [17]:
data['is_inconsistent_origin'] = ~np.isclose(data['oldbalanceOrg'] - data['newbalanceOrig'], data['amount'], atol=1.0)
data['is_inconsistent_dest'] = ~np.isclose(data['newbalanceDest'] - data['oldbalanceDest'], data['amount'], atol=1.0)

summary = (
    data.groupby('type')[['is_inconsistent_origin', 'is_inconsistent_dest']]
      .mean()
      .mul(100)
      .rename(columns={
          'is_inconsistent_origin': '%_inconsistent_origin',
          'is_inconsistent_dest': '%_inconsistent_dest'
      })
      .sort_values('%_inconsistent_origin', ascending=False)
)
summary


,%_inconsistent_origin,%_inconsistent_dest
type,,
CASH_IN,99.999786,99.999714
TRANSFER,95.237461,10.895294
CASH_OUT,88.541721,9.459531
PAYMENT,51.176879,99.995027
DEBIT,28.439371,8.314829


There are many inconsistencies but they should be analyzed separatedly because some type of movements only affect one either Orig or Dest. 

# Clean the data

### Inconsistent Column Names

In [18]:
df_fixed = data.copy()


### Standarize names

In [19]:
rename_map = {
        'oldbalanceOrg': 'oldbalance_orig',
        'newbalanceOrig': 'newbalance_orig',
        'oldbalanceDest': 'oldbalance_dest',
        'newbalanceDest': 'newbalance_dest',
        'nameOrig': 'name_orig',
        'nameDest': 'name_dest',
        'isFraud': 'is_fraud',
        'isFlaggedFraud': 'is_flagged_fraud'
    }
    
rename_map = {k: v for k, v in rename_map.items() if k in df_fixed.columns}

df_fixed = df_fixed.rename(columns=rename_map)
print(f"Renamed {len(rename_map)} columns")
for old, new in rename_map.items():
    print(f"{old} -> {new}")

else:
    print(" No columns to rename")

Renamed 8 columns
oldbalanceOrg -> oldbalance_orig
newbalanceOrig -> newbalance_orig
oldbalanceDest -> oldbalance_dest
newbalanceDest -> newbalance_dest
nameOrig -> name_orig
nameDest -> name_dest
isFraud -> is_fraud
isFlaggedFraud -> is_flagged_fraud
 No columns to rename


### Column isFraud with labels for Viz

In [20]:
df_fixed['is_fraud_label'] = df_fixed['is_fraud'].map({0: 'legitimate', 1: 'fraud'})

In [21]:
df_fixed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6362620 entries, 0 to 6362619
Data columns (total 15 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   step                    int64  
 1   type                    object 
 2   amount                  float64
 3   name_orig               object 
 4   oldbalance_orig         float64
 5   newbalance_orig         float64
 6   name_dest               object 
 7   oldbalance_dest         float64
 8   newbalance_dest         float64
 9   is_fraud                int64  
 10  is_flagged_fraud        int64  
 11  balance_diff            float64
 12  is_inconsistent_origin  bool   
 13  is_inconsistent_dest    bool   
 14  is_fraud_label          object 
dtypes: bool(2), float64(6), int64(3), object(4)
memory usage: 643.2+ MB


### Fix data types

I'm going to try to reduce the size of the dataframe by analyzing the types of data and in case that there are better options, change them.


Reasoning behind the following line:
- Step to int16: it it goes from 1 to 743. I assume that the df is created monthly and we will have the first approach using this time range. In case this changes and we start having more data to work with that needs a bigger range than int16 (-32,768 to 32,767 values) we should increase the data type to int32.

- Type, nameOrig, nameDest and isFraud to category: for better memory usage.

- Amount, oldbalanceOrig, newbalanceOrig, oldbalanceDest and newbalanceDest to float32: they all go from 0.0 to a maximum of 10^8. I assume that the all the balances cannot be much more bigger (i.e. assuming business knowledge related to type of transactions) so the range of float32 (±3.4 × 10^38) would be enough.

In [22]:
df_fixed[['step', 'is_inconsistent_origin', 'is_inconsistent_dest']] = df_fixed[['step', 'is_inconsistent_origin', 'is_inconsistent_dest']].astype('int16')
df_fixed[['type', 'name_orig', 'name_dest', 'is_fraud', 'is_flagged_fraud']] = df_fixed[['type', 'name_orig', 'name_dest', 'is_fraud', 'is_flagged_fraud']].astype('category')
df_fixed[['amount', 'oldbalance_orig', 'newbalance_orig', 'oldbalance_dest', 'newbalance_dest', 'balance_diff']] = df_fixed[['amount', 'oldbalance_orig', 'newbalance_orig', 'oldbalance_dest', 'newbalance_dest', 'balance_diff']].astype('float32')


In [23]:
df_fixed

,step,type,amount,name_orig,oldbalance_orig,newbalance_orig,name_dest,oldbalance_dest,newbalance_dest,is_fraud,is_flagged_fraud,balance_diff,is_inconsistent_origin,is_inconsistent_dest,is_fraud_label
0,1,PAYMENT,9.839640e+03,C1231006815,170136.000,160296.359375,M1979787155,0.000000e+00,0.000,0,0,9.839640e+03,0,1,legitimate
1,1,PAYMENT,1.864280e+03,C1666544295,21249.000,19384.720703,M2044282225,0.000000e+00,0.000,0,0,1.864280e+03,0,1,legitimate
2,1,TRANSFER,1.810000e+02,C1305486145,181.000,0.000000,C553264065,0.000000e+00,0.000,1,0,1.810000e+02,0,1,fraud
3,1,CASH_OUT,1.810000e+02,C840083671,181.000,0.000000,C38997010,2.118200e+04,0.000,1,0,1.810000e+02,0,1,fraud
4,1,PAYMENT,1.166814e+04,C2048537720,41554.000,29885.859375,M1230701703,0.000000e+00,0.000,0,0,1.166814e+04,0,1,legitimate
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6362615,743,CASH_OUT,3.396821e+05,C786484425,339682.125,0.000000,C776919290,0.000000e+00,339682.125,1,0,3.396821e+05,0,0,fraud
6362616,743,TRANSFER,6.311410e+06,C1529008245,6311409.500,0.000000,C1881841831,0.000000e+00,0.000,1,0,6.311410e+06,0,1,fraud
6362617,743,CASH_OUT,6.311410e+06,C1162922333,6311409.500,0.000000,C1365125890,6.848884e+04,6379898.000,1,0,6.311410e+06,0,0,fraud
6362618,743,TRANSFER,8.500025e+05,C1685995037,850002.500,0.000000,C2080388513,0.000000e+00,0.000,1,0,8.500025e+05,0,1,fraud


In [25]:
memory_usage = df_fixed.memory_usage(deep=True).sum()
print(f"Memory size: {memory_usage / (1024 ** 2):.2f} MB")

Memory size: 1432.87 MB


happy that I went from 1.65 GB 1.43 GB with 4 more columns added

### Save the dataset 

In [26]:
df_fixed.to_parquet('data/silver/df_fraud.parquet', index=False, engine='fastparquet')

In [27]:
df_fixed

,step,type,amount,name_orig,oldbalance_orig,newbalance_orig,name_dest,oldbalance_dest,newbalance_dest,is_fraud,is_flagged_fraud,balance_diff,is_inconsistent_origin,is_inconsistent_dest,is_fraud_label
0,1,PAYMENT,9.839640e+03,C1231006815,170136.000,160296.359375,M1979787155,0.000000e+00,0.000,0,0,9.839640e+03,0,1,legitimate
1,1,PAYMENT,1.864280e+03,C1666544295,21249.000,19384.720703,M2044282225,0.000000e+00,0.000,0,0,1.864280e+03,0,1,legitimate
2,1,TRANSFER,1.810000e+02,C1305486145,181.000,0.000000,C553264065,0.000000e+00,0.000,1,0,1.810000e+02,0,1,fraud
3,1,CASH_OUT,1.810000e+02,C840083671,181.000,0.000000,C38997010,2.118200e+04,0.000,1,0,1.810000e+02,0,1,fraud
4,1,PAYMENT,1.166814e+04,C2048537720,41554.000,29885.859375,M1230701703,0.000000e+00,0.000,0,0,1.166814e+04,0,1,legitimate
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6362615,743,CASH_OUT,3.396821e+05,C786484425,339682.125,0.000000,C776919290,0.000000e+00,339682.125,1,0,3.396821e+05,0,0,fraud
6362616,743,TRANSFER,6.311410e+06,C1529008245,6311409.500,0.000000,C1881841831,0.000000e+00,0.000,1,0,6.311410e+06,0,1,fraud
6362617,743,CASH_OUT,6.311410e+06,C1162922333,6311409.500,0.000000,C1365125890,6.848884e+04,6379898.000,1,0,6.311410e+06,0,0,fraud
6362618,743,TRANSFER,8.500025e+05,C1685995037,850002.500,0.000000,C2080388513,0.000000e+00,0.000,1,0,8.500025e+05,0,1,fraud
